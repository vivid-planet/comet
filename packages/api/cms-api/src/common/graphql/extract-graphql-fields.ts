import { GraphQLResolveInfo } from "graphql";
import { parseResolveInfo, ResolveTree } from "graphql-parse-resolve-info";

export function extractGraphqlFields(info: GraphQLResolveInfo, options: { root?: string } = {}): string[] {
    const resolveTree = parseResolveInfo(info) as ResolveTree;
    function treeToList(resolveTree: ResolveTree): string[] {
        const ret = [];
        for (const typeName in resolveTree.fieldsByTypeName) {
            const fields = resolveTree.fieldsByTypeName[typeName];
            for (const fieldName in fields) {
                ret.push(fields[fieldName].name);
                for (const i of treeToList(fields[fieldName])) {
                    ret.push(`${fields[fieldName].name}.${i}`);
                }
            }
        }
        return ret;
    }
    return treeToList(resolveTree).reduce((acc, i) => {
        if (!options.root) return [...acc, i];
        if (!i.startsWith(`${options.root}.`)) {
            return acc;
        }
        return [...acc, i.substring(options.root.length + 1)];
    }, [] as string[]);
}
