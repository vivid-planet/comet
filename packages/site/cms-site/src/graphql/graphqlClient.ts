import { GraphQLClient } from "graphql-request";

import { PreviewData } from "../sitePreview/SitePreviewApiHelper";

export function createGraphQLClient(url: string, previewData?: PreviewData): GraphQLClient {
    const { includeInvisibleBlocks, includeInvisiblePages, previewDamUrls } = {
        includeInvisiblePages: !!previewData,
        includeInvisibleBlocks: previewData && previewData.includeInvisible,
        previewDamUrls: !!previewData,
    };

    const headers: Record<string, string> = {};

    const includeInvisibleContentHeaderEntries: string[] = [];

    if (includeInvisiblePages) {
        includeInvisibleContentHeaderEntries.push("Pages:Unpublished");
    }

    if (includeInvisibleBlocks) {
        includeInvisibleContentHeaderEntries.push("Blocks:Invisible");
    }

    // tells api to send invisble content
    // authentication is required when this header is used
    if (includeInvisibleContentHeaderEntries.length > 0) {
        headers["x-include-invisible-content"] = includeInvisibleContentHeaderEntries.join(",");
    }

    // tells api to create preview image urls
    // authentication is required when this header is used
    if (previewDamUrls) {
        headers["x-preview-dam-urls"] = "1";
    }
    headers["x-relative-dam-urls"] = "1";

    return new GraphQLClient(url, {
        headers,
    });
}
