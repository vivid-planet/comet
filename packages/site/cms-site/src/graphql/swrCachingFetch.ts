const fetchCache: Record<string, { timestamp: number; ttl: number; swrTime: number; size: number; response: Response; updating?: true }> = {};

const defaultSwrTime = 60 * 5; //5 minutes
const defaultTtl = 60 * 5; //30 minutes
const maxCacheSize = 1024 * 1024 * 100; //100MB
const stats = {
    uncachable: 0,
    hit: 0,
    miss: 0,
    refresh: 0,
    timeout: 0,
};

interface SwrOptions {
    ttl?: number; //no cache if <= 0
    swrTime?: number; //no swr if <= 0
}

//cleanup cache every 5 minutes
// 1. remove all expired entries
// 2. if cache is too large, remove oldest entries
let cleanupStarted = false;
function avtivateCleanup() {
    if (cleanupStarted) return;
    cleanupStarted = true;

    setInterval(() => {
        let totalSize = 0;
        const now = new Date().getTime() / 1000;
        for (const key in fetchCache) {
            if (fetchCache[key].timestamp + fetchCache[key].ttl < now) {
                stats.timeout++;
                delete fetchCache[key];
            } else {
                totalSize += fetchCache[key].size;
            }
        }
        while (totalSize > maxCacheSize) {
            let oldestKey = "";
            let oldestTimestamp = now;
            for (const key in fetchCache) {
                if (fetchCache[key].timestamp < oldestTimestamp) {
                    oldestKey = key;
                    oldestTimestamp = fetchCache[key].timestamp;
                }
            }
            totalSize -= fetchCache[oldestKey].size;
            delete fetchCache[oldestKey];
        }
        // eslint-disable-next-line no-console
        console.log("swr fetch cache: ", Object.keys(fetchCache).length, "entries", `${Math.round(totalSize / 1024 / 1024)}MB`, stats);
    }, 1000 * 60 * 5);
}

export async function inMemorySwrCachingFetch(input: RequestInfo | URL, init?: RequestInit & SwrOptions): Promise<Response> {
    let cacheKey: string | undefined;
    if (init?.method?.toUpperCase() === "GET") {
        //cache all get requests
        cacheKey = input.toString();
    } else if (init?.body) {
        const body = JSON.parse(init.body.toString());
        if (body.query && body.variables) {
            //looks like a gql query, cache any method
            cacheKey = `${input.toString()}#${init.body.toString()}`;
        }
    }

    const ttl = init?.ttl ?? defaultTtl;
    const swrTime = init?.swrTime ?? defaultSwrTime;
    if (!cacheKey || ttl <= 0) {
        //uncachable, fetch live
        stats.uncachable++;
        return fetch(input, init);
    }

    const cachedResponse = fetchCache[cacheKey];
    if (cachedResponse) {
        if (ttl > 0 && new Date().getTime() / 1000 - cachedResponse.timestamp < cachedResponse.ttl) {
            if (new Date().getTime() / 1000 - cachedResponse.timestamp > cachedResponse.swrTime && !cachedResponse.updating) {
                //update cache in background
                cachedResponse.updating = true;
                stats.refresh++;
                fetch(input, init).then((response) => {
                    fetchCache[cacheKey as string] = {
                        timestamp: new Date().getTime() / 1000,
                        ttl,
                        swrTime,
                        size: response.headers.get("content-length") ? parseInt(response.headers.get("content-length") as string) : 5120, //random default of 5kb
                        response, //no clone, we don't return the response
                    };
                });
            }
            stats.hit++;
            return cachedResponse.response.clone();
        }
    }

    //fetch in foreground (still async)
    stats.miss++;
    const fetchPromise = fetch(input, init);
    const response = await fetchPromise;
    fetchCache[cacheKey] = {
        timestamp: new Date().getTime() / 1000,
        ttl,
        swrTime,
        size: response.headers.get("content-length") ? parseInt(response.headers.get("content-length") as string) : 5120, //random default of 5kb
        response: response.clone(),
    };
    avtivateCleanup(); //activate cleanup on first cache entry
    return fetchPromise;
}
