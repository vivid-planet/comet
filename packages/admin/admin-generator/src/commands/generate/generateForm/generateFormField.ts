import { type IntrospectionEnumType, type IntrospectionNamedTypeRef, type IntrospectionQuery } from "graphql";

import { type FormConfig, type FormFieldConfig, type GQLDocumentConfigMap } from "../generate-command";
import { camelCaseToHumanReadable } from "../utils/camelCaseToHumanReadable";
import { convertConfigImport } from "../utils/convertConfigImport";
import { type Imports } from "../utils/generateImportsCode";
import { isFieldOptional } from "../utils/isFieldOptional";
import { isGeneratorConfigCode, isGeneratorConfigImport } from "../utils/runtimeTypeGuards";
import { generateAsyncSelect } from "./asyncSelect/generateAsyncSelect";
import { buildFormFieldOptions } from "./formField/options";
import { type GenerateFieldsReturn } from "./generateFields";
import { type Prop } from "./generateForm";

export function generateFormField({
    gqlIntrospection,
    baseOutputFilename,
    config,
    formConfig,
    gqlType,
    namePrefix,
}: {
    gqlIntrospection: IntrospectionQuery;
    baseOutputFilename: string;
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    config: FormFieldConfig<any>;
    formFragmentName: string;
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    formConfig: FormConfig<any>;
    gqlType: string;
    namePrefix?: string;
}): GenerateFieldsReturn {
    if (config.type == "asyncSelect" || config.type == "asyncSelectFilter") {
        return generateAsyncSelect({ gqlIntrospection, baseOutputFilename, config, formConfig, gqlType, namePrefix });
    }
    const imports: Imports = [];
    const formProps: Prop[] = [];

    const {
        name,
        formattedMessageRootId,
        fieldLabel,
        startAdornment,
        endAdornment,
        imports: optionsImports,
    } = buildFormFieldOptions({ config, formConfig, gqlType, gqlIntrospection });
    imports.push(...optionsImports);

    let { introspectionFieldType } = buildFormFieldOptions({ config, formConfig, gqlType, gqlIntrospection });

    const nameWithPrefix = `${namePrefix ? `${namePrefix}.` : ``}${name}`;

    const rootGqlType = formConfig.gqlType;

    const dataRootName = rootGqlType[0].toLowerCase() + rootGqlType.substring(1); // TODO should probably be deteced via query

    const required = !isFieldOptional({ config, gqlIntrospection, gqlType });

    //TODO verify introspectionField.type is compatbile with config.type

    const endAdornmentWithLockIconProp = `endAdornment={<InputAdornment position="end"><Lock /></InputAdornment>}`;
    const readOnlyProps = `readOnly disabled`;
    const readOnlyPropsWithLock = `${readOnlyProps} ${endAdornmentWithLockIconProp}`;

    let formValuesConfig: GenerateFieldsReturn["formValuesConfig"] = [{}]; // FormFields should only contain one entry

    const gqlDocuments: GQLDocumentConfigMap = {};
    const hooksCode = "";
    let finalFormConfig: GenerateFieldsReturn["finalFormConfig"];

    let validateCode = "";
    if (config.validate) {
        if (isGeneratorConfigImport(config.validate)) {
            imports.push(convertConfigImport(config.validate));
            validateCode = `validate={${config.validate.name}}`;
        } else if (isGeneratorConfigCode(config.validate)) {
            validateCode = `validate={${config.validate.code}}`;
            imports.push(...config.validate.imports.map((imprt) => convertConfigImport(imprt)));
        }
    }

    let code = "";
    let formValueToGqlInputCode = "";
    let formFragmentFields = [name];
    if (config.type == "text") {
        const TextInputComponent = config.multiline ? "TextAreaField" : "TextField";
        code = `
        <${TextInputComponent}
            ${required ? "required" : ""}
            ${config.readOnly ? readOnlyPropsWithLock : ""}
            variant="horizontal"
            fullWidth
            name="${nameWithPrefix}"
            label={${fieldLabel}}
            ${config.startAdornment ? `startAdornment={<InputAdornment position="start">${startAdornment.adornmentString}</InputAdornment>}` : ""}
            ${config.endAdornment ? `endAdornment={<InputAdornment position="end">${endAdornment.adornmentString}</InputAdornment>}` : ""}
            ${
                config.helperText
                    ? `helperText={<FormattedMessage id=` +
                      `"${formattedMessageRootId}.${name}.helperText" ` +
                      `defaultMessage="${config.helperText}" />}`
                    : ""
            }
            ${validateCode}
        />`;
        if (!required && !config.readOnly) {
            formValueToGqlInputCode = `${name}: formValues.${name} ?? null,`;
        }
    } else if (config.type == "number") {
        code = `
            <NumberField
                ${required ? "required" : ""}
                ${config.readOnly ? readOnlyPropsWithLock : ""}
                variant="horizontal"
                fullWidth
                name="${nameWithPrefix}"
                label={${fieldLabel}}
                ${config.decimals ? `decimals={${config.decimals}}` : ""}
                ${config.startAdornment ? `startAdornment={<InputAdornment position="start">${startAdornment.adornmentString}</InputAdornment>}` : ""}
                ${config.endAdornment ? `endAdornment={<InputAdornment position="end">${endAdornment.adornmentString}</InputAdornment>}` : ""}
                ${
                    config.helperText
                        ? `helperText={<FormattedMessage id=` +
                          `"${formattedMessageRootId}.${name}.helperText" ` +
                          `defaultMessage="${config.helperText}" />}`
                        : ""
                }
                ${validateCode}
            />`;
        //TODO MUI suggest not using type=number https://mui.com/material-ui/react-text-field/#type-quot-number-quot
        let assignment = `parseFloat(formValues.${nameWithPrefix})`;
        if (isFieldOptional({ config, gqlIntrospection: gqlIntrospection, gqlType: gqlType })) {
            assignment = `formValues.${nameWithPrefix} ? ${assignment} : null`;
        }
        formValueToGqlInputCode = `${name}: ${assignment},`;

        let initializationAssignment = `String(data.${dataRootName}.${nameWithPrefix})`;
        if (!required) {
            initializationAssignment = `data.${dataRootName}.${nameWithPrefix} ? ${initializationAssignment} : undefined`;
        }
        formValuesConfig = [
            {
                omitFromFragmentType: name,
                typeCode: `${name}${!required ? `?` : ``}: string;`,
                initializationCode: `${name}: ${initializationAssignment}`,
            },
        ];
    } else if (config.type === "numberRange") {
        code = `
            <Field
                ${required ? "required" : ""}
                ${config.readOnly ? readOnlyPropsWithLock : ""}
                variant="horizontal"
                fullWidth
                name="${nameWithPrefix}"
                component={FinalFormRangeInput}
                label={${fieldLabel}}
                min={${config.minValue}}
                max={${config.maxValue}}
                ${config.disableSlider ? "disableSlider" : ""}
                ${config.startAdornment ? `startAdornment={<InputAdornment position="start">${startAdornment.adornmentString}</InputAdornment>}` : ""}
                ${config.endAdornment ? `endAdornment={<InputAdornment position="end">${endAdornment.adornmentString}</InputAdornment>}` : ""}
                ${
                    config.helperText
                        ? `helperText={<FormattedMessage id=` +
                          `"${formattedMessageRootId}.${name}.helperText" ` +
                          `defaultMessage="${config.helperText}" />}`
                        : ""
                }
                ${validateCode}
            />`;

        formFragmentFields = [`${name}.min`, `${name}.max`];
    } else if (config.type == "boolean") {
        code = `<CheckboxField
                        label={${fieldLabel}}
                        name="${nameWithPrefix}"
                        fullWidth
                        variant="horizontal"
                        ${config.readOnly ? readOnlyProps : ""}
                        ${
                            config.helperText
                                ? `helperText={<FormattedMessage id=` +
                                  `"${formattedMessageRootId}.${name}.helperText" ` +
                                  `defaultMessage="${config.helperText}" />}`
                                : ""
                        }
                        ${validateCode}
                    />`;
        formValuesConfig = [
            {
                defaultInitializationCode: `${name}: false`,
            },
        ];
    } else if (config.type == "date") {
        imports.push({
            name: "Future_DatePickerField",
            importPath: "@comet/admin",
        });
        code = `
            <Future_DatePickerField
                ${required ? "required" : ""}
                ${config.readOnly ? readOnlyPropsWithLock : ""}
                variant="horizontal"
                fullWidth
                name="${nameWithPrefix}"
                label={${fieldLabel}}
                ${config.startAdornment ? `startAdornment={<InputAdornment position="start">${startAdornment.adornmentString}</InputAdornment>}` : ""}
                ${config.endAdornment ? `endAdornment={<InputAdornment position="end">${endAdornment.adornmentString}</InputAdornment>}` : ""}
                ${
                    config.helperText
                        ? `helperText={<FormattedMessage id=` +
                          `"${formattedMessageRootId}.${name}.helperText" ` +
                          `defaultMessage="${config.helperText}" />}`
                        : ""
                }
                ${validateCode}
            />`;
        if (!required && !config.readOnly) {
            formValueToGqlInputCode = `${name}: formValues.${name} ?? null,`;
        }
    } else if (config.type == "dateTime") {
        code = `<DateTimeField
                ${required ? "required" : ""}
                ${config.readOnly ? readOnlyPropsWithLock : ""}
                variant="horizontal"
                fullWidth
                name="${nameWithPrefix}"
                label={${fieldLabel}}
                ${config.startAdornment ? `startAdornment={<InputAdornment position="start">${startAdornment.adornmentString}</InputAdornment>}` : ""}
                ${config.endAdornment ? `endAdornment={<InputAdornment position="end">${endAdornment.adornmentString}</InputAdornment>}` : ""}
                ${
                    config.helperText
                        ? `helperText={<FormattedMessage id=` +
                          `"${formattedMessageRootId}.${name}.helperText" ` +
                          `defaultMessage="${config.helperText}" />}`
                        : ""
                }
                ${validateCode}
            />`;
        formValuesConfig = [
            {
                initializationCode: `${name}: data.${dataRootName}.${nameWithPrefix} ? new Date(data.${dataRootName}.${nameWithPrefix}) : undefined`,
                omitFromFragmentType: name,
                typeCode: `${name}${!required ? "?" : ""}: Date${!required ? " | null" : ""};`,
            },
        ];
        if (!config.readOnly) {
            formValueToGqlInputCode = required
                ? `${name}: formValues.${name}.toISOString(),`
                : `${name}: formValues.${name} ? formValues.${name}.toISOString() : null,`;
        }
    } else if (config.type == "block") {
        code = `<Field name="${nameWithPrefix}" isEqual={isEqual} label={${fieldLabel}} variant="horizontal" fullWidth>
            {createFinalFormBlock(rootBlocks.${String(config.name)})}
        </Field>`;
        formValueToGqlInputCode = `${name}: rootBlocks.${name}.state2Output(formValues.${nameWithPrefix}),`;
        formValuesConfig = [
            {
                typeCode: `${name}: BlockState<typeof rootBlocks.${name}>;`,
                initializationCode: `${name}: rootBlocks.${name}.input2State(data.${dataRootName}.${nameWithPrefix})`,
                defaultInitializationCode: `${name}: rootBlocks.${name}.defaultValues()`,
            },
        ];
    } else if (config.type === "fileUpload") {
        const multiple = config.multiple || (typeof config.maxFiles === "number" && config.maxFiles > 1);
        code = `<FileUploadField name="${name}" label={${fieldLabel}}
            variant="horizontal"
            ${config.multiple ? "multiple" : ""}
            ${config.maxFiles ? `maxFiles={${config.maxFiles}}` : ""}
            ${config.maxFileSize ? `maxFileSize={${config.maxFileSize}}` : ""}
            ${config.readOnly ? `readOnly` : ""}
            ${config.layout ? `layout="${config.layout}"` : ""}
            ${config.accept ? `accept="${config.accept}"` : ""}
        />`;
        if (multiple) {
            formValueToGqlInputCode = `${name}: formValues.${name}?.map(({ id }) => id),`;
        } else {
            formValueToGqlInputCode = `${name}: formValues.${name} ? formValues.${name}.id : null,`;
        }
        formFragmentFields = [`${name} { ...${config.download ? "FinalFormFileUploadDownloadable" : "FinalFormFileUpload"} }`];
    } else if (config.type == "staticSelect") {
        const multiple = introspectionFieldType?.kind === "LIST";
        if (introspectionFieldType?.kind === "LIST") {
            introspectionFieldType =
                introspectionFieldType.ofType.kind === "NON_NULL" ? introspectionFieldType.ofType.ofType : introspectionFieldType.ofType;
        }

        const enumType = gqlIntrospection.__schema.types.find(
            (t) => t.kind === "ENUM" && t.name === (introspectionFieldType as IntrospectionNamedTypeRef).name,
        ) as IntrospectionEnumType | undefined;
        if (!enumType) throw new Error(`Enum type ${(introspectionFieldType as IntrospectionNamedTypeRef).name} not found for field ${name}`);

        const values = (config.values ? config.values : enumType.enumValues.map((i) => i.name)).map((value) => {
            if (typeof value === "string") {
                return {
                    value,
                    label: camelCaseToHumanReadable(value),
                };
            } else {
                return value;
            }
        });

        let inputType = config.inputType;
        if (!inputType) {
            if (!required || multiple) {
                // radio is not clearable, render always as select if not required
                // radio doesn't support multiple
                inputType = "select";
            } else {
                // auto select/radio based on number of values
                if (values.length <= 5) {
                    inputType = "radio";
                } else {
                    inputType = "select";
                }
            }
        }
        if (inputType === "radio" && multiple) {
            throw new Error(`${name}: inputType=radio doesn't support multiple`);
        }
        if (inputType === "radio" && !required) {
            throw new Error(`${name}: inputType=radio must be required as it doesn't support clearable`);
        }

        const renderAsRadio = inputType === "radio";
        if (renderAsRadio) {
            code = `<RadioGroupField
             ${required ? "required" : ""}
              variant="horizontal"
             fullWidth
             name="${nameWithPrefix}"
             label={${fieldLabel}}
             options={[
                  ${values
                      .map((value) => {
                          return `{
                                label: <FormattedMessage id="${formattedMessageRootId}.${name}.${
                                    value.value.charAt(0).toLowerCase() + value.value.slice(1)
                                }" defaultMessage="${value.label}" />,
                                value: "${value.value}",
                            }`;
                      })
                      .join(",")}
            ]}/>`;
        } else {
            imports.push({
                name: "SelectField",
                importPath: "@comet/admin",
            });
            code = `<SelectField
                            ${required ? "required" : ""}
                             fullWidth
                                variant={"horizontal"}
                                name="${nameWithPrefix}"
                                label={${fieldLabel}}
                                ${config.startAdornment ? `startAdornment={<InputAdornment position="start">${startAdornment.adornmentString}</InputAdornment>}` : ""}
                                ${
                                    config.helperText
                                        ? `helperText={<FormattedMessage id=` +
                                          `"${formattedMessageRootId}.${name}.helperText" ` +
                                          `defaultMessage="${config.helperText}" />}`
                                        : ""
                                }
                                ${validateCode}
                                ${config.readOnly ? readOnlyPropsWithLock : ""}
                                ${multiple ? "multiple" : ""}
                                options={[${values.map((value) => {
                                    const id = `${formattedMessageRootId}.${name}.${value.value.charAt(0).toLowerCase() + value.value.slice(1)}`;

                                    return `{
                                        value: "${value.value}",
                                        label: <FormattedMessage id="${id}" defaultMessage="${value.label}" />
                                    }`;
                                })}]}
                            />`;
        }
    } else {
        throw new Error(`Unsupported type`);
    }
    return {
        code,
        hooksCode,
        formValueToGqlInputCode,
        formFragmentFields,
        gqlDocuments,
        imports,
        formProps,
        formValuesConfig,
        finalFormConfig,
    };
}
