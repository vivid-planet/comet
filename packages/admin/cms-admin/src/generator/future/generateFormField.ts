import { IntrospectionEnumType, IntrospectionNamedTypeRef, IntrospectionObjectType, IntrospectionQuery } from "graphql";

import { FormConfig, FormFieldConfig, GeneratorReturn } from "./generator";
import { camelCaseToHumanReadable } from "./utils/camelCaseToHumanReadable";
import { Imports } from "./utils/generateImportsCode";

export function generateFormField(
    { gqlIntrospection }: { gqlIntrospection: IntrospectionQuery },
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    config: FormFieldConfig<any>,
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    formConfig: FormConfig<any>,
): GeneratorReturn & { imports: Imports } {
    const gqlType = formConfig.gqlType;
    const instanceGqlType = gqlType[0].toLowerCase() + gqlType.substring(1);

    const name = String(config.name);
    const label = config.label ?? camelCaseToHumanReadable(name);

    const introspectionObject = gqlIntrospection.__schema.types.find((type) => type.kind === "OBJECT" && type.name === gqlType) as
        | IntrospectionObjectType
        | undefined;
    if (!introspectionObject) throw new Error(`didn't find object ${gqlType} in gql introspection`);

    const introspectionField = introspectionObject.fields.find((field) => field.name === name);
    if (!introspectionField) throw new Error(`didn't find field ${name} in gql introspection type ${gqlType}`);
    const introspectionFieldType = introspectionField.type.kind === "NON_NULL" ? introspectionField.type.ofType : introspectionField.type;

    const requiredByIntrospection = introspectionField.type.kind == "NON_NULL";

    const required = config.required ?? requiredByIntrospection; //if undefined default to requiredByIntrospection

    //TODO verify introspectionField.type is compatbile with config.type

    const imports: Imports = [];

    let validateCode = "";
    if (config.validate) {
        let importPath = config.validate.import;
        if (importPath.startsWith("./")) {
            //go one level up as generated files are in generated subfolder
            importPath = `.${importPath}`;
        }
        imports.push({
            name: config.validate.name,
            importPath,
        });
        validateCode = `validate={${config.validate.name}}`;
    }

    let code = "";
    if (config.type == "text") {
        code = `
        <Field
            ${required ? "required" : ""}
            ${config.multiline ? "multiline" : ""}
            fullWidth
            name="${name}"
            component={FinalFormInput}
            label={<FormattedMessage id="${instanceGqlType}.${name}" defaultMessage="${label}" />}
            ${
                config.helperText
                    ? `helperText={<FormattedMessage id=` + `"${instanceGqlType}.${name}.helperText" ` + `defaultMessage="${config.helperText}" />}`
                    : ""
            }
            ${validateCode}
        />`;
    } else if (config.type == "number") {
        code = `
            <Field
                ${required ? "required" : ""}
                fullWidth
                name="${name}"
                component={FinalFormInput}
                type="number"
                label={<FormattedMessage id="${instanceGqlType}.${name}" defaultMessage="${label}" />}
                ${
                    config.helperText
                        ? `helperText={<FormattedMessage id=` +
                          `"${instanceGqlType}.${name}.helperText" ` +
                          `defaultMessage="${config.helperText}" />}`
                        : ""
                }
                ${validateCode}
            />`;
        //TODO MUI suggest not using type=number https://mui.com/material-ui/react-text-field/#type-quot-number-quot
    } else if (config.type == "boolean") {
        code = `<Field name="${name}" label="" type="checkbox" fullWidth ${validateCode}>
            {(props) => (
                <FormControlLabel
                    label={<FormattedMessage id="${instanceGqlType}.${name}" defaultMessage="${label}" />}
                    control={<FinalFormCheckbox {...props} />}
                    ${
                        config.helperText
                            ? `helperText={<FormattedMessage id=` +
                              `"${instanceGqlType}.${name}.helperText" ` +
                              `defaultMessage="${config.helperText}" />}`
                            : ""
                    }
                />
            )}
        </Field>`;
    } else if (config.type == "date") {
        code = `
            <Field
                ${required ? "required" : ""}
                fullWidth
                name="${name}"
                component={FinalFormDatePicker}
                label={<FormattedMessage id="${instanceGqlType}.${name}" defaultMessage="${label}" />}
                ${
                    config.helperText
                        ? `helperText={<FormattedMessage id=` +
                          `"${instanceGqlType}.${name}.helperText" ` +
                          `defaultMessage="${config.helperText}" />}`
                        : ""
                }
                ${validateCode}
            />`;
    } else if (config.type == "block") {
        imports.push({
            name: config.block.name,
            importPath: config.block.import,
        });
        code = `<Field name="${name}" isEqual={isEqual}>
            {createFinalFormBlock(${config.block.name})}
        </Field>`;
    } else if (config.type == "staticSelect") {
        if (config.values) {
            throw new Error("custom values for staticSelect is not yet supported"); // TODO add support
        }
        const enumType = gqlIntrospection.__schema.types.find(
            (t) => t.kind === "ENUM" && t.name === (introspectionFieldType as IntrospectionNamedTypeRef).name,
        ) as IntrospectionEnumType | undefined;
        if (!enumType) throw new Error(`Enum type ${(introspectionFieldType as IntrospectionNamedTypeRef).name} not found for field ${name}`);
        const values = enumType.enumValues.map((i) => i.name);
        code = `<Field
            fullWidth
            name="${name}"
            label={<FormattedMessage id="${instanceGqlType}.${name}" defaultMessage="${label}" />}>
            ${
                config.helperText
                    ? `helperText={<FormattedMessage id=` + `"${instanceGqlType}.${name}.helperText" ` + `defaultMessage="${config.helperText}" />}`
                    : ""
            }
            ${validateCode}
            {(props) => 
                <FinalFormSelect {...props}>
                ${values
                    .map((value) => {
                        const id = `${instanceGqlType}.${name}.${value.charAt(0).toLowerCase() + value.slice(1)}`;
                        const label = `<FormattedMessage id="${id}" defaultMessage="${camelCaseToHumanReadable(value)}" />`;
                        return `<MenuItem value="${value}">${label}</MenuItem>`;
                    })
                    .join("\n")}
                </FinalFormSelect>
            }
        </Field>`;
    } else {
        throw new Error(`Unsupported type: ${config.type}`);
    }
    return {
        code,
        gqlDocuments: {},
        imports,
    };
}
