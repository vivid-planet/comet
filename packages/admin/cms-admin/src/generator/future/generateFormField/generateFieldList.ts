import { IntrospectionField, IntrospectionQuery, IntrospectionType } from "graphql";

export function generateFieldListFromIntrospection(
    gqlIntrospection: IntrospectionQuery,
    type: string,
): { path: string; field: IntrospectionField }[] {
    return fieldListFromIntrospectionTypeRecursive(gqlIntrospection.__schema.types, type);
}

function fieldListFromIntrospectionTypeRecursive(
    types: readonly IntrospectionType[],
    type: string,
    parentPath?: string,
): { path: string; field: IntrospectionField }[] {
    const typeDef = types.find((typeDef) => typeDef.name === type);
    if (!typeDef || typeDef.kind !== "OBJECT") return [];

    return typeDef.fields.reduce<{ path: string; field: IntrospectionField }[]>((acc, field) => {
        const path = `${parentPath ? `${parentPath}.` : ""}${field.name}`;
        let type = field.type;
        if (type.kind === "NON_NULL") type = type.ofType;

        if (type.kind === "OBJECT") {
            acc.push({ path: path, field: field });
            acc.push(...fieldListFromIntrospectionTypeRecursive(types, type.name, path));
        } else {
            acc.push({ path: path, field: field });
        }
        return acc;
    }, []);
}
