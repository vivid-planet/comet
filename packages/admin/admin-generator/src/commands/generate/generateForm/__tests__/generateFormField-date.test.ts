import { buildSchema, introspectionFromSchema } from "graphql";

import type { FormConfig, FormFieldConfig } from "../../generate-command";
import { generateFormField } from "../generateFormField";

const schema = buildSchema(`
            """
            A local date string (i.e., with no associated timezone) in \`YYYY-MM-DD\` format, e.g. \`2020-01-01\`.
            """
            scalar LocalDate

            type Query {
                products: [Product]!
            } 

            type Product {
                id: ID!
                availableSince: LocalDate
            }
          
            type Mutation {
                createProduct(input: ProductInput!): Product!
                updateProduct(id: ID!, input: ProductInput!): Product!
            }

            input ProductInput {
                title: String
                availableSince: LocalDate = null
            }
        `);

type GQLProduct = {
    __typename?: "Product";
    id: string;
    availableSince: string | null;
};

describe("generateFormField - datePicker", () => {
    it("should generate simple date field with LocalType scalar", async () => {
        const fieldConfig: FormFieldConfig<GQLProduct> = {
            type: "date",
            name: "availableSince",
        };

        const formConfig: FormConfig<GQLProduct> = {
            type: "form",
            gqlType: "Product",
            fields: [fieldConfig],
        };

        const introspection = introspectionFromSchema(schema);

        const formOutput = generateFormField({
            gqlIntrospection: introspection,
            baseOutputFilename: "ProductForm",
            formFragmentName: "ProductFormFragment",
            config: fieldConfig,
            formConfig,
            gqlType: "Product",
        });
        expect(formOutput.code).toMatchSnapshot();
        expect(formOutput.formValuesConfig).toMatchSnapshot();
    });
});
