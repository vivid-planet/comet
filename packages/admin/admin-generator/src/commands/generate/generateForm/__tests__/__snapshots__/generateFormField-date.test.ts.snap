// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generateFormField - datePicker should generate simple date field with LocalType scalar 1`] = `
"
            <Future_DatePickerField
                
                
                variant="horizontal"
                fullWidth
                name="availableSince"
                label={<FormattedMessage id="product.availableSince" defaultMessage="Available Since" />}
                
                
                
                
            />"
`;

exports[`generateFormField - datePicker should generate simple date field with LocalType scalar 2`] = `
[
  {
    "fieldName": "availableSince",
    "formValueToGqlInputCode": "formValues.availableSince ?? null",
  },
]
`;
